Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -113,25 +155,82 @@ extension HTTP1ToRawGRPCServerCodec: ChannelInboundHandler {
 extension HTTP1ToRawGRPCServerCodec: ChannelOutboundHandler {
   public typealias OutboundIn = RawGRPCServerResponsePart
   public typealias OutboundOut = HTTPServerResponsePart
-  
+
   public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
     let responsePart = self.unwrapOutboundIn(data)
     switch responsePart {
-    case .headers(let headers):
-      //! FIXME: Should return a different version if we want to support pPRC.
-      ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: .init(major: 2, minor: 0), status: .ok, headers: headers))), promise: promise)
+    case .headers:
+      var headers = HTTPHeaders()
+      let version: HTTPVersion
+      if let contentType = contentType {
+        headers.add(name: "content-type", value: contentType.rawValue)
+        switch contentType {
+        case .text, .web:
+          version = .init(major: 1, minor: 1)
+        case .binary:
+          version = .init(major: 2, minor: 0)
+        }
+      } else {
+        version = .init(major: 2, minor: 0)
+      }
+
+      if contentType == .text {
+        responseTextBuffer = ctx.channel.allocator.buffer(capacity: 0)
+      }
+
+      ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: version, status: .ok, headers: headers))), promise: promise)
     case .message(var messageBytes):
       // Write out a length-delimited message payload. See `channelRead` fpor the corresponding format.
       var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.readableBytes + 5)
       responseBuffer.write(integer: Int8(0))  // Compression flag: no compression
       responseBuffer.write(integer: UInt32(messageBytes.readableBytes))
       responseBuffer.write(buffer: &messageBytes)
-      ctx.write(self.wrapOutboundOut(.body(.byteBuffer(responseBuffer))), promise: promise)
+
+      if contentType == .text {
+        precondition(responseTextBuffer != nil)
+        // Store the response into an independent buffer. We can't return the message directly as
+        // it needs to be aggregated with all the responses plus the trailers, in order to have
+        // the base64 response properly encoded in a single byte stream.
+        responseTextBuffer!.write(buffer: &responseBuffer)
```

Comment:
I understand the necessity to cache all responses to ensure a well-formed base64 stream, but doesn't buffering all of them defeat the purpose of streaming? How about instead trying to send all but the last `length - (length % 3)` bytes, as those should always result in four well-formed base64 bytes? Please at least consider that when refactoring the base64 handling out of this class in a separate PR.

Reply:
Isn't that the issue with HTTP1.1, that there is no proper server streaming? I may be wrong though. if we did do that, we'd still have an issue with some streams, as if there are 2 bytes missing for a full base64 frame, we'd have to wait until the trailers or the next messages to arrive before sending it, otherwise we'd be sending incomplete messages. Perhaps that's ok? not sure

Reply:
HTTP1.1 already has [chunking](https://en.wikipedia.org/wiki/Chunked_transfer_encoding), which is essentially server streaming. Also, I think most gRPC clients should be capable of waiting for the next chunk mid-message, as the gRPC spec says that clients should be chunking-agnostic. I'd assume the same applies for HTTP1.1 — the client just receives a byte stream and waits for a full message to arrive before trying to decode it.

Would you mind adding a `FIXME` for this?